### ISSUE TITLE
Infinite Loop in spark

### ISSUE THREAD

**Issue by zhangjunqiang** [eclipse/deeplearning4j#4448]
#### Issue Description
Recently, I got a problem with dl4j on spark on yarn.
After several epoch（it is random）, a worknode on yarn will enter a infinite loop. like this:
<code>
org.bytedeco.javacpp.openblas.cblas_sgemm(Native Method)
org.nd4j.linalg.cpu.nativecpu.blas.CpuLevel3.sgemm(CpuLevel3.java:38)
org.nd4j.linalg.api.blas.impl.BaseLevel3.gemm(BaseLevel3.java:84)
org.nd4j.linalg.factory.Nd4j.gemm(Nd4j.java:930)
org.deeplearning4j.nn.layers.recurrent.LSTMHelpers.backpropGradientHelper(LSTMHelpers.java:624)
org.deeplearning4j.nn.layers.recurrent.GravesLSTM.backpropGradientHelper(GravesLSTM.java:102)
org.deeplearning4j.nn.layers.recurrent.GravesLSTM.tbpttBackpropGradient(GravesLSTM.java:79)
org.deeplearning4j.nn.multilayer.MultiLayerNetwork.truncatedBPTTGradient(MultiLayerNetwork.java:1541)
org.deeplearning4j.nn.multilayer.MultiLayerNetwork.computeGradientAndScore(MultiLayerNetwork.java:2219)
org.deeplearning4j.optimize.solvers.BaseOptimizer.gradientAndScore(BaseOptimizer.java:174)
org.deeplearning4j.optimize.solvers.StochasticGradientDescent.optimize(StochasticGradientDescent.java:60)
org.deeplearning4j.optimize.Solver.optimize(Solver.java:53)
org.deeplearning4j.nn.multilayer.MultiLayerNetwork.doTruncatedBPTT(MultiLayerNetwork.java:1468)
org.deeplearning4j.nn.multilayer.MultiLayerNetwork.fit(MultiLayerNetwork.java:1824)
org.deeplearning4j.spark.impl.paramavg.ParameterAveragingTrainingWorker.processMinibatch(ParameterAveragingTrainingWorker.java:179)
org.deeplearning4j.spark.impl.paramavg.ParameterAveragingTrainingWorker.processMinibatch(ParameterAveragingTrainingWorker.java:41)
org.deeplearning4j.spark.api.worker.ExecuteWorkerFlatMapAdapter.call(ExecuteWorkerFlatMap.java:125)
org.deeplearning4j.spark.api.worker.ExecuteWorkerFlatMapAdapter.call(ExecuteWorkerFlatMap.java:42)
org.deeplearning4j.spark.api.worker.ExecuteWorkerPathFlatMapAdapter.call(ExecuteWorkerPathFlatMap.java:71)
org.deeplearning4j.spark.api.worker.ExecuteWorkerPathFlatMapAdapter.call(ExecuteWorkerPathFlatMap.java:38)
org.datavec.spark.transform.BaseFlatMapFunctionAdaptee.call(BaseFlatMapFunctionAdaptee.java:24)
org.apache.spark.api.java.JavaRDDLike$$anonfun$fn$4$1.apply(JavaRDDLike.scala:153)
org.apache.spark.api.java.JavaRDDLike$$anonfun$fn$4$1.apply(JavaRDDLike.scala:153)
org.apache.spark.rdd.RDD$$anonfun$mapPartitions$1$$anonfun$apply$23.apply(RDD.scala:797)
org.apache.spark.rdd.RDD$$anonfun$mapPartitions$1$$anonfun$apply$23.apply(RDD.scala:797)
org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:38)
org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:323)
org.apache.spark.rdd.RDD.iterator(RDD.scala:287)
org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:38)
org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:323)
org.apache.spark.rdd.RDD.iterator(RDD.scala:287)
org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:38)
org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:323)
org.apache.spark.rdd.RDD.iterator(RDD.scala:287)
org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:96)
org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:53)
org.apache.spark.scheduler.Task.run(Task.scala:108)
org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:335)
java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
</code>

I don't know what's wrong with my code,  Can anyone help me？
#### Version Information
* Spark Version : 2.2.0
* Spark On Yarn
* My dependency:
  &lt;dependency>
            &lt;groupId>org.deeplearning4j&lt;/groupId>
            &lt;artifactId>dl4j-spark_2.11&lt;/artifactId>
            &lt;version>0.9.1_spark_2&lt;/version>
  &lt;/dependency>
* Platform:Linux

#### My Code
<code>
 val lstmLayerSize = 128 
val networkConfig = new NeuralNetConfiguration
      .Builder()
      .optimizationAlgo(OptimizationAlgorithm.STOCHASTIC_GRADIENT_DESCENT)
      .iterations(1).learningRate(0.001)
      .seed(12345)
      .weightInit(WeightInit.XAVIER)
      .updater(Updater.RMSPROP)
      .list
      .layer(0, new GravesLSTM.Builder().nIn(featureDimension).nOut(lstmLayerSize).activation(Activation.TANH).build)
      .layer(1, new GravesLSTM.Builder().nIn(lstmLayerSize).nOut(lstmLayerSize).activation(Activation.TANH).build)
      .layer(2, new GravesLSTM.Builder().nIn(lstmLayerSize).nOut(lstmLayerSize).activation(Activation.TANH).build)
//      .layer(3, new GravesLSTM.Builder().nIn(lstmLayerSize).nOut(lstmLayerSize).activation(Activation.TANH).build)
      .layer(3, new RnnOutputLayer.Builder(LossFunctions.LossFunction.MCXENT).activation(Activation.SOFTMAX).nIn(lstmLayerSize).nOut(2).build)
      .backpropType(BackpropType.TruncatedBPTT)
      .tBPTTForwardLength(timeSeqLength).tBPTTBackwardLength(timeSeqLength)
      .pretrain(false)
      .backprop(true)
      .build

    val net = new MultiLayerNetwork(networkConfig)
    net.setListeners(new ScoreIterationListener(1))

    val tm = new ParameterAveragingTrainingMaster
        .Builder(1)
        .batchSizePerWorker(16)
        .averagingFrequency(5)
        .workerPrefetchNumBatches(2)
        .saveUpdater(true)
        .rddTrainingApproach(RDDTrainingApproach.Export)
        .build

    val sparkNet = new SparkDl4jMultiLayer(sparkSession.sparkContext, networkConfig, tm)
  val sparkNet = new SparkDl4jMultiLayer(sparkSession.sparkContext, networkConfig, tm)

    val numEpochs = 30
    var i = 1
    for(_ <- 1 to numEpochs) {
      sparkNet.fit(trainDataSet)
    }
</code>

**Comment by raver119** [comment]
@saudet Looks like yet another openblas issue?

**Comment by saudet** [comment]
@zhangjunqiang Are you saying it never returns from `cblas_sgemm()`? 

In any case, might be worth to try setting `OMP_NUM_THREADS=1`, yes.

**Comment by zhangjunqiang** [comment]
@saudet  Yes， it's always running and never return.

**Comment by saudet** [comment]
@zhangjunqiang Please let us know if this happens even with `OMP_NUM_THREADS=1`! Thanks

**Comment by zhangjunqiang** [comment]
@saudet 
I set the omp num like this.
<code>
NativeOpsHolder.getInstance().getDeviceNativeOps.setOmpNumThreads(1)
</code>
Still the same result.

**Comment by saudet** [comment]
This won't change the number of threads used by OpenBLAS. If you can't set an environment variable, try to call `Nd4j.factory().blas().setMaxThread(1)` instead.

**Comment by zhangjunqiang** [comment]
@saudet I did as you said. But still the same. orz

**Comment by saudet** [comment]
So maybe it's a bug in the optimization. There are ways to work around that as well: https://deeplearning4j.org/native#troubleshooting-the-cpu-backend

**Comment by zhangjunqiang** [comment]
@saudet How to set  BLAS environment? Could u help me? I can not modify the online environment directly. Can I set these environment in the jar? Thank u very much!

**Comment by zhangjunqiang** [comment]
@saudet Should i install openblas for every yarn node ?

**Comment by saudet** [comment]
There are different ways to set environment variables with Spark, some are explained here:
http://spark.apache.org/docs/latest/configuration.html

**Comment by saudet** [comment]
Another option is to install MKL in the PATH. It will get picked up automatically by ND4J, and is pretty much always better than OpenBLAS.

**Comment by Beili** [comment]
sasdas

**Comment by Beili** [comment]
It worked when I decrease num-executors  and set  `OMP_NUM_THREADS=1 `
I set  the omp num in spark-submit commands:
`--conf spark.executorEnv.OMP_NUM_THREADS=1`
